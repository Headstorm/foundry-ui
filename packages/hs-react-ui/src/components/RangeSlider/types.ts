import { SubcomponentPropsType } from '../commonTypes';

export type ValueProp = {
  value: number;
  label?: string | number | Node;
  color?: string;
};

export type ContainerProps = {
  showDomainLabels?: boolean;
  hasHandleLabels?: boolean;
  disabled: boolean;
  beingDragged: boolean;
};

export type HandleProps = {
  beingDragged?: boolean;
  color: string;
};

export type HandleLabelProps = { velocity?: number };

export type SelectedRangeProps = { min: number; max: number; selectedRange: number[] };

export type DomainLabelProps = { position: 'left' | 'right' };

export type RangeSliderProps = {
  StyledContainer?: any;
  StyledDragHandle?: any;
  StyledHandleLabel?: any;
  StyledSlideRail?: any;
  StyledSelectedRangeRail?: any;
  StyledDomainLabel?: any;
  StyledMarker?: any;
  StyledMarkerLabel?: any;
  containerProps?: SubcomponentPropsType;
  dragHandleProps?: SubcomponentPropsType;
  handleLabelProps?: SubcomponentPropsType;
  slideRailProps?: SubcomponentPropsType;
  selectedRangeRailProps?: SubcomponentPropsType;
  domainLabelProps?: SubcomponentPropsType;
  markerProps?: SubcomponentPropsType;
  markerLabelProps?: SubcomponentPropsType;

  showDomainLabels?: boolean;
  showSelectedRange?: boolean;

  motionBlur?: boolean;
  springOnRelease?: boolean;
  debounceInterval?: number;
  axisLock?: 'x' | 'y' | '';
  onDrag?: (val: number) => void;
  disabled?: boolean;
  min: number;
  max: number;
  values?: number[] | ValueProp[];
  testId?: string;
  markers?: number[] | ValueProp[];
};
