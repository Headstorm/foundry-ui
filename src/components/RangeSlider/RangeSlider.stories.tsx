import React, { useState, useEffect, forwardRef } from 'react';
import styled from 'styled-components';
import { Story, Meta } from '@storybook/react';
import { readableColor, toColorString } from 'polished';

import { action } from '@storybook/addon-actions';
import { withFoundryContext } from '../../../.storybook/decorators';
import fonts from '../../enums/fonts';
import colors from '../../enums/colors';
import RangeSlider, { SlideRail } from './RangeSlider';
import { RangeSliderProps, ValueProp } from './types';

import Card from '../Card';

const Row = styled.div`
  ${fonts.body}
  display: flex;
  flex-flow: row nowrap;
  align-items: center;
  padding: 0.5em;
  min-width: 20em;
  width: 90%;
  margin: 0 auto;
`;

const ColorPreview = styled.div`
  ${fonts.body}

  height: 12rem;
  width: 34rem;
  transition: color 0.5s;
  font-size: 5rem;
  display: flex;
  justify-content: center;
  align-items: center;
  text-transform: uppercase;
  font-weight: 900;
`;

const skillLabels = [
  'WhatJS?',
  "I've used it...",
  "We're aight",
  'I love it',
  'I dream in React',
  'I am React',
];

const skillColors = ['red', 'orangered', 'orange', 'goldenrod', 'yellowgreen', 'forestgreen'];

const StyledSlideRail = styled(SlideRail)`
  filter: grayscale(0.5) brightness(1.3);
  border: 0.5px solid ${colors.grayLight};
  height: 0.5em;
  border-radius: 0.25rem;
  background-image: linear-gradient(to right, ${skillColors.join(', ')});
`;

type DefaultProps = Omit<RangeSliderProps, 'markers'> & {
  value: number;
  'use marker labels': boolean;
  markers: string;
};

export const Default: Story<DefaultProps> = ({
  value,
  markers,
  'use marker labels': markerLabels,
  disabled,
  showDomainLabels,
  showSelectedRange,
  min,
  max,
  dragHandleAttachment,
  readonly,
  debounceInterval,
  animated,
}: DefaultProps) => {
  const [val, setVal] = useState(value);

  useEffect(() => {
    setVal(value);
  }, [value]);

  const markersSelection = markers;
  const markersArray: Array<number | ValueProp> = [];
  if (markersSelection === 'all values') {
    for (let i = min; i <= max; i++) {
      markersArray.push(markerLabels ? { value: i, label: `${i}` } : i);
    }
  } else if (markersSelection === 'middle value') {
    const midpoint = (min + max) / 2;
    markersArray.push(markerLabels ? { value: midpoint, label: `${midpoint}` } : midpoint);
  }

  return (
    <Row>
      <RangeSlider
        disabled={disabled}
        readonly={readonly}
        showDomainLabels={showDomainLabels}
        showSelectedRange={showSelectedRange}
        animated={animated}
        min={min}
        max={max}
        debounceInterval={debounceInterval}
        onChange={newVal => {
          action('onChange')(newVal);
          setVal(Math.round(newVal));
        }}
        dragHandleAttachment={dragHandleAttachment}
        values={[{ value: val, label: val }]}
        markers={markersArray as RangeSliderProps['markers']}
      />
    </Row>
  );
};
Default.args = {
  value: 1,
  min: 0,
  max: 5,
  markers: 'none',
  'use marker labels': false,
  disabled: false,
  readonly: false,
  showDomainLabels: false,
  showHandleLabels: true,
  showSelectedRange: true,
  dragHandleAttachment: 'value',
  debounceInterval: 10,
  animated: true,
};

type RatingProps = Omit<RangeSliderProps, 'markers'> & {
  value: number;
};

export const Rating: Story<RatingProps> = ({
  value,
  disabled,
  showDomainLabels,
  showSelectedRange,
  min,
  max,
}: RatingProps) => {
  const [val, setVal] = useState(value);

  return (
    <Row>
      <span>ReactJS:&nbsp;&nbsp;&nbsp;&nbsp;</span>
      <RangeSlider
        StyledSlideRail={StyledSlideRail}
        disabled={disabled}
        showDomainLabels={showDomainLabels}
        showSelectedRange={showSelectedRange}
        min={min}
        max={max}
        onChange={newVal => {
          action('onChange')(newVal);
          setVal(Math.round(newVal));
        }}
        values={[
          {
            value: val,
            label: skillLabels[Math.round(val)],
            color: skillColors[Math.round(val)],
          },
        ]}
      />
    </Row>
  );
};
Rating.args = {
  value: 0,
  disabled: false,
  showDomainLabels: false,
  showSelectedRange: false,
  min: 0,
  max: 5,
};

interface ColorPickerProps {
  hue: number;
  lightness: number;
  saturation: number;
  disabled: boolean;
  showDomainLabels: boolean;
  debounceInterval: number;
}

export const ColorPicker: Story<ColorPickerProps> = ({
  hue,
  lightness,
  saturation,
  disabled,
  showDomainLabels,
  debounceInterval,
}: ColorPickerProps) => {
  const [hue_, setHue] = useState(hue);
  const [sat, setSat] = useState(saturation);
  const [light, setLight] = useState(lightness);

  useEffect(() => {
    setHue(hue_);
  }, [hue_]);

  useEffect(() => {
    setSat(saturation);
  }, [saturation]);

  useEffect(() => {
    setLight(lightness);
  }, [lightness]);

  const allHues = Array.from({ length: 360 }, (_, i) => i).map(
    num => `hsl(${num}, ${sat}%, ${light}%)`,
  );
  const allSats = [`hsl(${hue_}, 0%, ${light}%)`, `hsl(${hue_}, 100%, ${light}%`];
  const allLights = [`hsl(${hue_}, ${sat}%, 10%)`, `hsl(${hue_}, ${sat}%, 90%)`];

  return (
    <Card
      elevation={2}
      StyledHeader={Card.NoPaddingHeader}
      header={
        <ColorPreview
          style={{
            backgroundColor: `hsl(${hue_},${sat}%,${light}%)`,
            color: readableColor(`hsl(${hue_},${sat}%,${light}%)`),
          }}
        >
          {toColorString({
            hue: hue_,
            saturation: sat / 100,
            lightness: light / 100,
          })}
        </ColorPreview>
      }
    >
      <Row>
        <span>H:&nbsp;&nbsp;&nbsp;</span>
        <RangeSlider
          StyledSlideRail={forwardRef((props, ref) => (
            <SlideRail
              ref={ref as React.RefObject<HTMLDivElement>}
              {...props}
              style={{
                backgroundImage: `linear-gradient(to right, ${allHues.join(', ')})`,
              }}
            />
          ))}
          disabled={disabled}
          showDomainLabels={showDomainLabels}
          showSelectedRange={false}
          min={0}
          max={360}
          debounceInterval={debounceInterval}
          onDebounceChange={val => {
            action('onDebounceChange hue')(val);
            setHue(Math.round(val));
          }}
          values={[
            {
              value: hue_,
              label: hue_,
              color: colors.grayLight,
            },
          ]}
        />
      </Row>
      <Row>
        <span>S:&nbsp;&nbsp;&nbsp;</span>
        <RangeSlider
          StyledSlideRail={forwardRef((props, ref) => (
            <SlideRail
              ref={ref as React.RefObject<HTMLDivElement>}
              {...props}
              style={{
                backgroundImage: `linear-gradient(to right, ${allSats.join(', ')})`,
              }}
            />
          ))}
          min={0}
          max={100}
          debounceInterval={debounceInterval}
          onDebounceChange={val => {
            action('onDebounceChange saturation')(val);
            setSat(Math.round(val));
          }}
          showDomainLabels={false}
          showSelectedRange={false}
          values={[
            {
              value: sat,
              label: sat,
              color: colors.grayLight,
            },
          ]}
        />
      </Row>
      <Row>
        <span>L:&nbsp;&nbsp;&nbsp;</span>
        <RangeSlider
          StyledSlideRail={forwardRef((props, ref) => (
            <SlideRail
              // eslint-disable-next-line @typescript-eslint/ban-ts-comment
              // @ts-ignore
              ref={ref}
              {...props}
              style={{
                backgroundImage: `linear-gradient(to right, ${allLights.join(', ')})`,
              }}
            />
          ))}
          min={0}
          max={100}
          debounceInterval={debounceInterval}
          onDebounceChange={val => {
            action('onDebounceChange lightness')(val);
            setLight(Math.round(val));
          }}
          showDomainLabels={false}
          showSelectedRange={false}
          values={[
            {
              value: light,
              label: light,
              color: colors.grayLight,
            },
          ]}
        />
      </Row>
    </Card>
  );
};
ColorPicker.args = {
  hue: 0,
  saturation: 50,
  lightness: 50,
  disabled: false,
  showDomainLabels: false,
  debounceInterval: 10,
};

export default {
  title: 'RangeSlider',
  argTypes: {
    value: {
      control: {
        type: 'range',
        min: 0,
        max: 5,
        step: 1,
      },
    },
    min: {
      control: { type: 'range', min: -10, max: 10, step: 1 },
    },
    max: {
      control: {
        type: 'range',
        min: -10,
        max: 10,
        step: 1,
      },
    },
    debounceInterval: {
      control: {
        type: 'range',
        min: 0,
        max: 100,
        step: 1,
      },
    },
    markers: {
      options: ['none', 'all values', 'middle value'],
      control: {
        type: 'radio',
      },
    },
    hue: {
      control: {
        type: 'range',
        min: 0,
        max: 360,
        step: 1,
      },
    },
    saturation: {
      control: {
        type: 'range',
        min: 0,
        max: 100,
        step: 1,
      },
    },
    lightness: {
      control: {
        type: 'range',
        min: 0,
        max: 100,
        step: 1,
      },
    },
    dragHandleAttachment: {
      control: {
        type: 'radio',
      },
      options: ['mouse', 'value'],
    },
  },
  decorators: [withFoundryContext],
  parameters: {
    design: {
      type: 'figma',
      url: 'https://www.figma.com/file/3r2G00brulOwr9j7F6JF59/Generic-UI-Style?node-id=126%3A2',
    },
  },
} as Meta;
