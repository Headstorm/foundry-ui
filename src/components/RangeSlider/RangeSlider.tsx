import React, { useEffect, useRef, useMemo, useState, useCallback } from 'react';
import styled from 'styled-components';
import debounce from 'lodash/debounce';

import { useSpring, a } from '@react-spring/web';
import { useDrag } from 'react-use-gesture';
import useMeasure from 'react-use-measure';
import { ResizeObserver } from '@juggle/resize-observer';

import fonts from '../../enums/fonts';
import { clamp } from '../../utils/math';
import { mergeRefs } from '../../utils/refs';

import {
  ValueProp,
  ContainerProps,
  HandleProps,
  HandleLabelProps,
  RangeSliderProps,
  DomainLabelProps,
} from './types';
import { useAccessibilityPreferences, useAnalytics, useTheme } from '../../context';
import { StyledBaseDiv } from '../../htmlElements';

export const Container = styled.div`
  ${({ showDomainLabels, hasHandleLabels, disabled, beingDragged = false }: ContainerProps) => `
    display: flex;
    position: relative;
    height: 1rem;
    width: 100%;

    ${fonts.body}

    user-select: none;

    ${beingDragged ? 'cursor: grabbing;' : ''}

    transition: filter .1s;

    ${
      disabled
        ? `
        filter: grayscale(1) contrast(.5) brightness(1.2);
        pointer-events: none;
      `
        : ''
    }

    

    ${
      showDomainLabels
        ? `
        top: -.5rem;
        margin-top: 1rem;
      `
        : ''
    };

    ${
      hasHandleLabels
        ? `
      top: -.75rem;
      margin-top: 1.5rem;
    `
        : ''
    };
  `}
`;

export const DragHandle = styled(a.div)`
  ${({ $beingDragged = false, color, $readonly }: HandleProps) => {
    const { colors } = useTheme();
    const handleColor = color || colors.primary;
    return `
      position: absolute;
      
      width: 1rem;
      height: 1rem;
      align-self: center;
      left: -.5rem;

      background-color: ${handleColor};
      color: ${handleColor};
      border: .125rem solid ${colors.background};
      border-radius: 50%;
      
      touch-action: none;
      filter: url(#blur);
      cursor: ${$beingDragged ? 'grabbing' : 'grab'};
      cursor: ${$readonly ? 'default' : ''};
      z-index: 2;
    `;
  }}
`;

export const HandleLabel = styled.div`
  ${({ velocity = 0 }: HandleLabelProps) => {
    const { colors } = useTheme();
    return `
      position: absolute;
      bottom: 100%;
      left: 50%;
      transform: translateX(-50%) rotate(${clamp(velocity, -45, 45)}deg);
      background-color: ${colors.background};
      border-radius: 4px;
      font-weight: bold;
      white-space: nowrap;
      pointer-events: none;
      z-index: 2;
    `;
  }}
`;

export const SlideRail = styled.div`
  ${() => {
    const { colors } = useTheme();
    return `
      position: absolute;
      width: 100%;
      height: 0.25rem;

      overflow: hidden;
      align-self: center;

      border-radius: 0.125rem;
      background-color: ${colors.grayXlight};
    `;
  }}
`;

export const SelectedRangeRail = styled(a.div)`
  ${() => {
    const { colors } = useTheme(); // TODO: don't force the color to be primary
    return `
      position: absolute;
      top: 0%;
      height: 100%;

      background-color: ${colors.primary};
    `;
  }}
`;

export const DomainLabel = styled.div`
  ${({ position }: DomainLabelProps) => {
    const { colors } = useTheme();
    return `
      position: absolute;
      bottom: 100%;
      ${position}: 0rem;
      color: ${colors.grayMedium};
      font-size: .5rem;
    `;
  }}
`;

export const Marker = styled(StyledBaseDiv)`
  ${({ sliderPosition = 0 }) => {
    const { colors } = useTheme();
    return `
      position: absolute;
      text-align: center;
      display: flex;
      justify-content: center;
      height: 1rem;
      width: 2px;
      left: ${sliderPosition}px;
      background-color: ${colors.grayLight};
    `;
  }}
`;
export const MarkerLabel = styled(StyledBaseDiv)`
  ${({ color }) => {
    const { colors } = useTheme();
    return `
    position: absolute;
    bottom: 100%;
    white-space: nowrap;
    font-size: .375rem;
    color: ${color || colors.grayLight};
  `;
  }}
`;

export const RangeSlider = ({
  StyledContainer = Container,
  StyledDragHandle = DragHandle,
  StyledHandleLabel = HandleLabel,
  StyledSlideRail = SlideRail,
  StyledSelectedRangeRail = SelectedRangeRail,
  StyledDomainLabel = DomainLabel,
  StyledMarker = Marker,
  StyledMarkerLabel = MarkerLabel,

  containerProps = {},
  dragHandleProps = {},
  handleLabelProps = {},
  slideRailProps = {},
  selectedRangeRailProps = {},
  domainLabelProps = {},
  markerProps = {},
  markerLabelProps = {},

  containerRef,
  dragHandleRef,
  slideRailRef,
  handleLabelRef,
  selectedRangeRailRef,
  domainLabelRef,
  markerRef,
  markerLabelRef,

  showDomainLabels = true,
  showSelectedRange = true,
  showHandleLabels = true,

  springOnRelease,
  animated = true,

  debounceInterval = 8,
  onDrag,
  onChange,
  onDebounceChange,
  onRelease,

  disabled = false,
  readonly = false,
  min,
  max,
  values,
  markers = [],
  testId,
  dragHandleAttachment = 'mouse',
}: RangeSliderProps): JSX.Element | null => {
  if (onDrag) {
    // eslint-disable-next-line no-console
    console.warn(
      'From FoundryUI RangerSlider: onDrag callback is deprecated. Instead, use onChange or onDebounceChange.',
    );
  }
  if (springOnRelease !== undefined) {
    animated = springOnRelease;
    // eslint-disable-next-line no-console
    console.warn(
      'From FoundryUI RangeSlider: springOnRelease is deprecated. Instead, use animated.',
    );
  }

  const { prefersReducedMotion } = useAccessibilityPreferences();
  const isInitializing = useRef(true);

  const debouncedOnChange = useRef(
    debounce(newVal => {
      if (onDrag) onDrag(newVal);
      if (onDebounceChange) onDebounceChange(newVal);
    }, debounceInterval),
  ).current;

  const debouncedOnRelease = useRef(
    // wait an extra ms. onRelease should be called after onChange
    debounce(newVal => onRelease && onRelease(newVal), debounceInterval + 1),
  ).current;

  /** Convert passed-in `number` values into `ValueProps` */
  const processVal = (val: number | ValueProp): ValueProp =>
    typeof val === 'number' ? { value: val, label: undefined, color: undefined } : val;

  const processedValues: Array<ValueProp> = useMemo(
    // eslint-disable-next-line @typescript-eslint/ban-ts-comment
    // @ts-ignore expression not callable
    () => values?.map(processVal) ?? [],
    [values],
  );

  const processedMarkers: Array<ValueProp> = useMemo(
    // eslint-disable-next-line @typescript-eslint/ban-ts-comment
    // @ts-ignore expression not callable
    () => markers?.map(processVal) ?? [],
    [markers],
  );

  const hasHandleLabels = useMemo(
    () => processedValues?.some(val => val.label !== null && val.label !== undefined),
    [processedValues],
  );

  const domain = max - min;

  const handleEventWithAnalytics = useAnalytics();

  const handleDrag = useCallback(
    (newVal: number) => {
      if (readonly) return;

      handleEventWithAnalytics(
        'RangeSlider',
        () => {
          if (onChange) onChange(newVal);
          debouncedOnChange(newVal);
        },
        'onDrag',
        { type: 'onDrag', newVal },
        containerProps,
      );
    },

    [readonly, handleEventWithAnalytics, containerProps, onChange, debouncedOnChange],
  );

  // set the drag value asynchronously at a lower frequency for better performance
  const valueBuffer = useRef(0);

  // keep track of which handle is being dragged (if any)
  const [draggedHandle, setDraggedHandle] = useState(-1);
  // get the bounding box of the slider

  const [ref, sliderBounds] = useMeasure({ polyfill: ResizeObserver });

  const pixelPositions = processedValues.map(val => {
    return (val.value / domain) * sliderBounds.width;
  });

  // get the x offset and an animation setter function
  const [{ dragHandleX }, springRef] = useSpring(() => ({
    to: { dragHandleX: 0 },
    friction: 13,
    tension: 100,
  }));

  const handleSlideRailClick = useCallback(
    (e: React.MouseEvent) => {
      // Avoiding using another ref here to reduce overhead
      const pixelPosition = e.clientX;
      const positionOnRail = pixelPosition - sliderBounds.left;
      const railPositionRatio = positionOnRail / sliderBounds.width;
      const clickedValue = railPositionRatio * domain;

      // variables to find the closest handle
      let closestVal: number | ValueProp | undefined;
      let smallestDifference: number;

      // Find the closest handle
      processedValues.forEach(val => {
        const finalVal = typeof val === 'number' ? val : val.value;
        // Get the absolute value of the difference
        const difference = Math.abs(clickedValue - finalVal);
        if (smallestDifference !== undefined && difference < smallestDifference) {
          closestVal = val;
          smallestDifference = difference;
        } else if (smallestDifference === undefined) {
          closestVal = val;
          smallestDifference = difference;
        }
      });

      if (closestVal) {
        if (onDrag) onDrag(clickedValue);
        if (onChange) onChange(clickedValue);
        if (onDebounceChange) onDebounceChange(clickedValue);
        if (onRelease) onRelease(clickedValue);

        if (slideRailProps.onMouseDown && typeof slideRailProps.onMouseDown === 'function') {
          e.persist();
          slideRailProps.onMouseDown(e);
        }
      }
    },
    [
      slideRailProps,
      sliderBounds,
      onChange,
      onDebounceChange,
      onRelease,
      onDrag,
      domain,
      processedValues,
    ],
  );

  const handleSlideRailClickWithAnalytics = (e: any) => {
    if (readonly) return;
    handleEventWithAnalytics('RangeSlider', handleSlideRailClick, 'onClick', e, containerProps);
  };

  const bind = useDrag(
    ({ down: isDragging, movement: [deltaX] }) => {
      if (readonly) return;

      const delta = (deltaX / sliderBounds.width) * domain;
      valueBuffer.current = clamp(delta, min, max);
      setDraggedHandle(isDragging ? 0 : -1);
      handleDrag(valueBuffer.current);

      if (dragHandleAttachment === 'mouse') {
        if (isDragging) {
          // constantly follow mouse during drag
          springRef.start({
            dragHandleX: deltaX,
            immediate: true,
          });
        } else {
          // after drag release, spring to value
          springRef.start({
            dragHandleX: pixelPositions[0],
            immediate: prefersReducedMotion || !animated,
          });
        }
      }
    },
    {
      initial: [(pixelPositions ?? [0])[0], 0],
      threshold: 1,
      bounds: {
        left: 0,
        right: sliderBounds.width + 4,
        top: -8,
        bottom: sliderBounds.height / 2 + 8,
      },
      rubberband: 0.1,
    },
  );

  // Once sliderBounds are read, set initial position
  useEffect(() => {
    if (isInitializing.current && sliderBounds.width) {
      springRef.start({
        dragHandleX: pixelPositions[0],
        immediate: true,
        onResolve: () => {
          isInitializing.current = false;
        },
      });
    }
  }, [springRef, sliderBounds, isInitializing, pixelPositions]);

  // For snap to value, listen to changes in value and always animate to value. Also listens to clicks
  useEffect(() => {
    if (dragHandleAttachment === 'value' || draggedHandle === -1) {
      springRef.start({
        dragHandleX: pixelPositions[0],
        immediate: prefersReducedMotion || !animated,
      });
    }
  }, [
    dragHandleAttachment,
    springRef,
    pixelPositions,
    prefersReducedMotion,
    sliderBounds,
    animated,
    draggedHandle,
  ]);

  // Dispose of debounce timers
  useEffect(() => {
    return () => {
      debouncedOnChange.cancel();
      debouncedOnRelease.cancel();
    };
  }, [debouncedOnChange, debouncedOnRelease]);

  return (
    <StyledContainer
      data-test-id={['hs-ui-range-slider', testId].join('-')}
      disabled={disabled}
      readonly={readonly}
      hasHandleLabels={hasHandleLabels}
      showHandleLabels={showHandleLabels}
      showDomainLabels={showDomainLabels}
      ref={containerRef}
      {...containerProps}
    >
      <StyledSlideRail
        ref={mergeRefs<HTMLDivElement>([slideRailRef, ref])}
        {...slideRailProps}
        onMouseDown={handleSlideRailClickWithAnalytics}
      >
        {showSelectedRange && values && (
          <StyledSelectedRangeRail
            min={min}
            max={max}
            values={processedValues}
            style={{
              width: dragHandleX,
            }}
            ref={selectedRangeRailRef}
            {...selectedRangeRailProps}
          />
        )}
      </StyledSlideRail>

      {showDomainLabels && (
        <>
          <StyledDomainLabel position="left" ref={domainLabelRef} {...domainLabelProps}>
            {min}
          </StyledDomainLabel>
          <StyledDomainLabel position="right" {...domainLabelProps}>
            {max}
          </StyledDomainLabel>
        </>
      )}

      {processedValues.map(({ value, color, label }, i) => (
        <StyledDragHandle
          // eslint-disable-next-line react/jsx-props-no-spreading
          {...bind()}
          draggable={false}
          $beingDragged={i === draggedHandle}
          style={{ x: dragHandleX }}
          color={color}
          // eslint-disable-next-line react/no-array-index-key
          key={`handle${i}`}
          ref={dragHandleRef}
          onMouseUp={() => debouncedOnRelease(value)}
          $readonly={readonly}
          {...dragHandleProps}
        >
          {showHandleLabels && (
            <StyledHandleLabel value={value} ref={handleLabelRef} {...handleLabelProps}>
              {label}
            </StyledHandleLabel>
          )}
        </StyledDragHandle>
      ))}

      {processedMarkers.map(({ value, color, label }) => (
        <StyledMarker
          key={`marker-${value}`}
          id={`marker-${value}`}
          sliderPosition={(value / domain) * sliderBounds.width}
          ref={markerRef}
          {...markerProps}
        >
          <StyledMarkerLabel color={color} ref={markerLabelRef} {...markerLabelProps}>
            {label}
          </StyledMarkerLabel>
        </StyledMarker>
      ))}
    </StyledContainer>
  );
};

RangeSlider.Container = Container;
RangeSlider.DragHandle = DragHandle;
RangeSlider.HandleLabel = HandleLabel;
RangeSlider.SlideRail = SlideRail;
RangeSlider.SelectedRangeRail = SelectedRangeRail;
RangeSlider.DomainLabel = DomainLabel;
RangeSlider.Marker = Marker;
RangeSlider.MarkerLabel = MarkerLabel;

export default RangeSlider;
