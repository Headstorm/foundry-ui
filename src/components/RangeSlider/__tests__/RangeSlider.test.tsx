import React from 'react';
import { render, fireEvent, waitFor, configure } from '@testing-library/react';
import { axe, toHaveNoViolations } from 'jest-axe';
import RangeSlider from '../RangeSlider';

expect.extend(toHaveNoViolations);
configure({ testIdAttribute: 'data-test-id' });

const testId = 'hs-ui-range-slider-unit-test';

describe('RangeSlider', () => {
  it('renders', async () => {
    const { container, getByTestId } = render(
      <RangeSlider min={0} max={10} testId="unit-test" values={[0]} />,
    );

    await waitFor(() => getByTestId(testId));

    expect(container).toMatchSnapshot();
  });

  it('renders', async () => {
    const { container, getByTestId } = render(
      <RangeSlider values={[3]} markers={[3]} min={0} max={10} testId="unit-test" />,
    );

    await waitFor(() => getByTestId(testId));
    fireEvent.mouseDown(getByTestId(testId));
    expect(container).toMatchSnapshot();
  });

  // it('Fire handler on click event ', async () => {
  //   const spy = jest.fn();
  //   const { getByTestId } = render(<RangeSlider slideRailProps={{onMouseDown: spy}} min={0} max={10} testId="unit-test" />);
  //   await waitFor(() => getByTestId(testId));
  //   fireEvent.mouseDown(getByTestId(testId));
  //   expect(spy).toHaveBeenCalled();
  // });

  describe('Accessibility Tests', () => {
    it('Should pass accessibility test with default props', async () => {
      const component = <RangeSlider min={0} max={10} testId={testId} values={[1]} />;
      const { container } = render(component);
      const results = await axe(container);
      expect(results).toHaveNoViolations();
    });
  });
  describe('Ref tests', () => {
    it('containerRef.current should exist', async () => {
      const ref = React.createRef<HTMLDivElement>();
      const { getByTestId } = render(
        <RangeSlider min={0} max={10} containerRef={ref} testId="unit-test" values={[1]} />,
      );
      await waitFor(() => getByTestId(testId));
      expect(ref.current instanceof HTMLDivElement).toBeTruthy();
    });
  });
});
