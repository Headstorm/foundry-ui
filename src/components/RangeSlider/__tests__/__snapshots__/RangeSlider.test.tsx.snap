// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`RangeSlider renders 1`] = `
.c0 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  position: relative;
  height: 1rem;
  width: 100%;
  font-family: Arial,Roboto,sans-serif;
  -webkit-user-select: none;
  -moz-user-select: none;
  -ms-user-select: none;
  user-select: none;
  -webkit-transition: filter .1s;
  transition: filter .1s;
  top: -.5rem;
  margin-top: 1rem;
}

.c5 {
  position: absolute;
  width: 1rem;
  height: 1rem;
  -webkit-align-self: center;
  -ms-flex-item-align: center;
  align-self: center;
  left: -.5rem;
  background-color: #DA4200;
  color: #DA4200;
  border: .125rem solid #fff;
  border-radius: 50%;
  touch-action: none;
  -webkit-filter: url(#blur);
  filter: url(#blur);
  cursor: -webkit-grab;
  cursor: -moz-grab;
  cursor: grab;
  z-index: 2;
}

.c6 {
  position: absolute;
  bottom: 100%;
  left: 50%;
  -webkit-transform: translateX(-50%) rotate(0deg);
  -ms-transform: translateX(-50%) rotate(0deg);
  transform: translateX(-50%) rotate(0deg);
  background-color: #fff;
  border-radius: 4px;
  font-weight: bold;
  white-space: nowrap;
  pointer-events: none;
  z-index: 2;
}

.c1 {
  position: absolute;
  width: 100%;
  height: 0.25rem;
  overflow: hidden;
  -webkit-align-self: center;
  -ms-flex-item-align: center;
  align-self: center;
  border-radius: 0.125rem;
  background-color: #D5D5D5;
}

.c2 {
  position: absolute;
  top: 0%;
  height: 100%;
  background-color: #DA4200;
}

.c3 {
  position: absolute;
  bottom: 100%;
  left: 0rem;
  color: #51575D;
  font-size: .5rem;
}

.c4 {
  position: absolute;
  bottom: 100%;
  right: 0rem;
  color: #51575D;
  font-size: .5rem;
}

<div>
  <div
    class="c0"
    data-test-id="hs-ui-range-slider-unit-test"
  >
    <div
      class="c1"
    >
      <div
        class="c2"
        max="10"
        min="0"
        style="width: 0px;"
        values="[object Object]"
      />
    </div>
    <div
      class="c3"
    >
      0
    </div>
    <div
      class="c4"
    >
      10
    </div>
    <div
      class="c5"
      draggable="false"
      style="transform: none;"
    >
      <div
        class="c6"
        value="0"
      />
    </div>
  </div>
</div>
`;

exports[`RangeSlider renders 2`] = `
.c0 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  position: relative;
  height: 1rem;
  width: 100%;
  font-family: Arial,Roboto,sans-serif;
  -webkit-user-select: none;
  -moz-user-select: none;
  -ms-user-select: none;
  user-select: none;
  -webkit-transition: filter .1s;
  transition: filter .1s;
  top: -.5rem;
  margin-top: 1rem;
}

.c5 {
  position: absolute;
  width: 1rem;
  height: 1rem;
  -webkit-align-self: center;
  -ms-flex-item-align: center;
  align-self: center;
  left: -.5rem;
  background-color: #DA4200;
  color: #DA4200;
  border: .125rem solid #fff;
  border-radius: 50%;
  touch-action: none;
  -webkit-filter: url(#blur);
  filter: url(#blur);
  cursor: -webkit-grab;
  cursor: -moz-grab;
  cursor: grab;
  z-index: 2;
}

.c6 {
  position: absolute;
  bottom: 100%;
  left: 50%;
  -webkit-transform: translateX(-50%) rotate(0deg);
  -ms-transform: translateX(-50%) rotate(0deg);
  transform: translateX(-50%) rotate(0deg);
  background-color: #fff;
  border-radius: 4px;
  font-weight: bold;
  white-space: nowrap;
  pointer-events: none;
  z-index: 2;
}

.c1 {
  position: absolute;
  width: 100%;
  height: 0.25rem;
  overflow: hidden;
  -webkit-align-self: center;
  -ms-flex-item-align: center;
  align-self: center;
  border-radius: 0.125rem;
  background-color: #D5D5D5;
}

.c2 {
  position: absolute;
  top: 0%;
  height: 100%;
  background-color: #DA4200;
}

.c3 {
  position: absolute;
  bottom: 100%;
  left: 0rem;
  color: #51575D;
  font-size: .5rem;
}

.c4 {
  position: absolute;
  bottom: 100%;
  right: 0rem;
  color: #51575D;
  font-size: .5rem;
}

.c7 {
  position: absolute;
  text-align: center;
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-box-pack: center;
  -webkit-justify-content: center;
  -ms-flex-pack: center;
  justify-content: center;
  height: 1rem;
  width: 2px;
  left: 0px;
  background-color: #9FA2A5;
}

.c8 {
  position: absolute;
  bottom: 100%;
  white-space: nowrap;
  font-size: .375rem;
  color: #9FA2A5;
}

<div>
  <div
    class="c0"
    data-test-id="hs-ui-range-slider-unit-test"
  >
    <div
      class="c1"
    >
      <div
        class="c2"
        max="10"
        min="0"
        style="width: 0px;"
        values="[object Object]"
      />
    </div>
    <div
      class="c3"
    >
      0
    </div>
    <div
      class="c4"
    >
      10
    </div>
    <div
      class="c5"
      draggable="false"
      style="transform: none;"
    >
      <div
        class="c6"
        value="3"
      />
    </div>
    <div
      class=" c7"
      id="marker-3"
    >
      <div
        class=" c8"
      />
    </div>
  </div>
</div>
`;
